### PR TITLE
OS-1061: Add scripts to generate introspection schema

### PR DESCRIPTION
**Pull request by josemarinas**
## Description

Currently the introspection schema json was generatated from the schema.graphql file. However this wasnt the correct form of doing it because it added some types that broke the subgraph deployment and the whole schema wasent available.

Subraph generates an eschema on top of your `schema.graphql` and we dont have access to this schema before deploying the subgraph itself, so we needed to query the subgraph to get the full schema after it was deployed.

We had 2 options here either calling the deployecd subgraph in alchemy or generating it from a locally deployed subgraph. The firs one wasnt the best option because we needed to deploy the whole subgraph to have the full schema, deploying locally we can get the same result without depending on external services.

What was done:

- The `docker-compose` got updated, all the services got updated and the `hardhat-node` built fro the `hardhat.Dockerfile` got simplified to just deploy a HH node
- An script that feches the local subgraph and saves the result to a json file
- An script to do all this operations in one go

Some minor twaks were done in order to make this work.

Task ID: [OS-1061](https://aragonassociation.atlassian.net/browse/OS-1061)

## Type of change

See the framework lifecylce in `packages/contracts/docs/framework-lifecycle` to decide what kind of change this pull request is.

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)

## Checklist:

- [x] I have selected the correct base branch.
- [x] I have performed a self-review of my own code.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] My changes generate no new warnings.
- [x] Any dependent changes have been merged and published in downstream modules.
- [x] I ran all tests with success and extended them if necessary.
- [ ] I have updated the `CHANGELOG.md` file in the root folder.
- [ ] I have updated the `DEPLOYMENT_CHECKLIST` file in the root folder.
- [ ] I have updated the `UPDATE_CHECKLIST` file in the root folder.
- [ ] I have updated the Subgraph and added a QA URL to the description of this PR.

[OS-1061]: https://aragonassociation.atlassian.net/browse/OS-1061?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: packages/subgraph/docker/docker-compose.yml [hunk 1]
```diff
@@ -1,52 +1,75 @@
...
 services:
-  hardhat:
-    build:
-      context: ../../../
-      dockerfile: packages/subgraph/docker/hardhat.Dockerfile
+  hardhat-node:
+    build: 
+      context: .
+      dockerfile: hardhat.Dockerfile 
+    container_name: ${PROJECT_NAME}-hardhat-node
```
Comment (Rekard0) [inline packages/subgraph/docker/docker-compose.yml:7]: can we have default for all variables?

File: packages/subgraph/docker/hardhat/package.json [hunk 2]
```diff
@@ -0,0 +1,6 @@
+{
+  "name": "hardhat-project",
```
Comment (Rekard0) [inline packages/subgraph/docker/hardhat/package.json:2]: Can we use the hardhard project from packages/contracts with minimum effort ? if not create a follow up task

File: packages/subgraph/manifest/data/localhost.json [hunk 3]
```diff
@@ -5,18 +5,18 @@
     "DAORegistry": {
       "name": "DAORegistry",
       "address": "0x9A9f2CCfdE556A7E9Ff0848998Aa4a0CFD8863AE",
-      "startBlock": 18
+      "startBlock": 0
```
Comment (Rekard0) [inline packages/subgraph/manifest/data/localhost.json:8]: is the addres here are dummy addresses ? if yes, can we use address 1,2,3...etc?

File: packages/subgraph/scripts/generate-introspection-json.sh [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+    echo "env variables are not set properly, exiting..."
+    exit -1
+fi
+
+# Exit script as soon as a command fails.
+set -o errexit
+# Up docker
+docker-compose -f docker/docker-compose.yml up -d
+# Wait for containers to start
+sleep 10
```
Comment (Rekard0) [inline packages/subgraph/scripts/generate-introspection-json.sh:19]: may be this delay is not stable and not enough,
what we could do is to loop with delay 2 seconds, and once the docker is up then continue with the rest of the code

File: packages/subgraph/package.json [hunk 5]
```diff
@@ -14,7 +14,7 @@
     "deploy": "scripts/deploy-subgraph.sh",
     "create:local": "graph create aragon/aragon-core-goerli --node http://localhost:8020",
     "deploy:local": "LOCAL=true scripts/deploy-subgraph.sh",
-    "generate:introspection": "graphql-codegen && cd docs && prettier schema-introspection.json --write",
+    "generate:introspection": "scripts/generate-introspection-json.sh",
```
Comment (Rekard0) [inline packages/subgraph/package.json:17]: perhaps we need to create 2 action here one for partial & one for full introsepction

File: packages/subgraph/scripts/generate-introspection-json.sh [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+if [ -z "$NETWORK_NAME" ] || [ -z "$SUBGRAPH_NAME" ] || [ -z "$GRAPH_KEY" ] || [ -z "$SUBGRAPH_VERSION" ]
+then
+    echo "env variables are not set properly, exiting..."
+    exit -1
+fi
+
+# Exit script as soon as a command fails.
+set -o errexit
+# Up docker
+docker-compose -f docker/docker-compose.yml up -d
```
Comment (Rekard0) [inline packages/subgraph/scripts/generate-introspection-json.sh:17]: update to docker compose v2

File: packages/subgraph/docker/docker-compose.yml [hunk 7]
```diff
@@ -1,52 +1,75 @@
...
+        aliases:
+          - ${POSTGRES_HOST:-postgres}
+
+  graph-node:
+    image: graphprotocol/graph-node:${GRAPH_NODE_VERSION:-latest}
+    container_name: ${PROJECT_NAME:-local-subgraph}-graph-node
+    platform: linux/amd64
+    depends_on:
+      - ipfs
+      - postgres
```
Comment (banasa44) [inline packages/subgraph/docker/docker-compose.yml:51]: why don't wait for HH-node too?

**Review by banasa44 — DISMISSED** [review]

File: packages/subgraph/docker/.env.example [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+GRAPH_NODE_SERVER_PORT=8000
+GRAPH_NODE_SERVER_WS_PORT=8001
+GRAPH_NODE_ADMIN_PORT=8020
+GRAPH_NODE_INDEX_NODE_SERVER_PORT=8030
+GRAPH_NODE_METRICS_PORT=8040
+GRAPH_NODE_HOST=graph-node
+GRAPH_NODE_VOLUME_DATA=./data/graph-node
+# HARDHAT
+HARDHAT_PORT=8545
+HARDHAT_HOST=hardhat
```
Comment (Rekard0) [inline packages/subgraph/docker/.env.example:25]:
```suggestion
HARDHAT_HOST=hardhat

```

File: packages/subgraph/manifest/data/localhost.json [hunk 9]
```diff
@@ -4,20 +4,20 @@
...
         "name": "PluginSetupProcessor",
-        "address": "0x59b670e9fA9D0A427751Af201D676719a970857b",
-        "startBlock": 22
+        "address": "0x0000000000000000000000000000000000000003",
+        "startBlock": 0
       }
     ]
   }
-}
+}
```
Comment (Rekard0) [inline packages/subgraph/manifest/data/localhost.json:23]:
```suggestion
}

```

File: packages/subgraph/package.json [hunk 10]
```diff
@@ -40,4 +41,4 @@
   "dependencies": {
     "@aragon/osx-commons-subgraph": "^0.0.4"
   }
-}
+}
```
Comment (Rekard0) [inline packages/subgraph/package.json:44]:
```suggestion
}

```

File: packages/subgraph/scripts/generate-introspection-json.sh [hunk 11]
```diff
@@ -0,0 +1,43 @@
...
+# deploy subgraph
+LOCAL=true ./scripts/deploy-subgraph.sh
+# make introspection json
+FILENAME="docs/schema-introspection.json"
+GRAPH_URL="http://localhost:8000/subgraphs/name/$SUBGRAPH_NAME-$NETWORK_NAME"
+echo "Fetching introspection query from $GRAPH_URL"
+node ./scripts/fetch-introspection-query.js $GRAPH_URL $FILENAME
+prettier $FILENAME --write
+# down docker
+docker compose -f docker/docker-compose.yml down
```
Comment (Rekard0) [inline packages/subgraph/scripts/generate-introspection-json.sh:43]:
```suggestion
docker compose -f docker/docker-compose.yml down

```

**Review by banasa44 — APPROVED** [review]

**Review by Rekard0 — APPROVED** [review]

**Pull request merged**